warehouse/query-core/src/main/java/datawave/query/language/parser/jexl/JexlNodeSet.java

```diff
@@ -0,0 +1,185 @@
+package datawave.query.language.parser.jexl;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import datawave.query.jexl.JexlASTHelper;
+import org.apache.commons.jexl2.parser.JexlNode;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utility class that implements the {@link Set} interface for use with a collection of Jexl nodes.
+ *
+ * Element uniqueness is computed using {@link JexlASTHelper#nodeToKey(JexlNode)} for keys.
+ *
+ * Important Note: The set is considered invalid if any of the underlying Jexl nodes change.
+ */
+public class JexlNodeSet implements Set<JexlNode> {
```

Comment:
We could achieve the same functionality here if we just overrode `equals()` and `hashCode()` in `JexlNode`. Then we could use any of the Java collections we wanted (like HashSet for this case). It'd be less code and also enable us to switch collection implementations easier if we ever needed to.

Reply:
We could, and that's probably the best long term solution. However, we're working with apache jexl nodes, so we would have to either submit a PR against that project or override the class.

The danger we run into is when the JexlNode represents a large tree and it becomes expensive to compute the node key. With this implementation of JexlNodeSet, we can limit where we perform this potentially expensive operation (i.e., only pay the cost when we need it vs. everywhere all the time).